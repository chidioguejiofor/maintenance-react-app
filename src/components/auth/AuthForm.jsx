import { Form, Container, Header, Icon, Message } from 'semantic-ui-react';
import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import TopNav from '../Header';

const AuthForm = ({
  formType, handleSubmit, success, messageTitle,
  loading, handleChange, inputs, errorMessages,
  links
}) => (
  <div id="auth-form">
    <TopNav />
    <Container>
      <div className="form">
        <Form
          loading={loading}
          onSubmit={handleSubmit}
          error={!!errorMessages.length}
          success={success}
        >

          <Header as="h2" icon textAlign="center">
            <Icon name="user circle" />
            <Header.Content>{formType}</Header.Content>
          </Header>
          <Message
            header={messageTitle}
            error={!!errorMessages.length}
            success={success}
            list={errorMessages}
            hidden={!success && !errorMessages.length}
          />
          {inputs.map((input, index) => (
            <Form.Input
              key={index}
              label={input.placeholder}
              placeholder={input.placeholder}
              name={input.name}
              value={input.value}
              onChange={handleChange}
              type={input.type}
            />
          ))}
          <Form.Button
            disabled={success}
            positive
            content="Submit"
          />
          {links.map((linkObj, index) => (
            <span key={index} className="link">
              <Link to={linkObj.to}>{linkObj.caption}</Link>
            </span>

          ))}
        </Form>
      </div>

    </Container>
  </div>


);

AuthForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired,
  inputs: PropTypes.arrayOf(PropTypes.objectOf(PropTypes.string)).isRequired,
  formType: PropTypes.string,
  errorMessages: PropTypes.arrayOf(PropTypes.string),
  success: PropTypes.bool,
  messageTitle: PropTypes.string,
  loading: PropTypes.bool,
  links: PropTypes.arrayOf(PropTypes.object),
};
AuthForm.defaultProps = {
  formType: 'Sign Up',
  errorMessages: [],
  success: true,
  messageTitle: '',
  loading: false,
  links: [],
};
export default AuthForm;
