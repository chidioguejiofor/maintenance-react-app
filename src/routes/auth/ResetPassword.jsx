import React from 'react';


const ResetPassword = () => (
  <h1>This is the reset password</h1>
);

export default ResetPassword;
