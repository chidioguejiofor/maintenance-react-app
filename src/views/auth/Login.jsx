import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Form from '../../components/Form';
import { login as loginAction,
  engineerLogin as engineerLoginAction } from '../../actions/authAction';
import Header from '../../components/Header';
/**
 *contains logic for the LoginPage.
 */
export class Login extends Component {
  /**
   * this creates a new instance of Login class.
   * The super constructor is called here and the state is initialized
   * @param {object} props contains the props sent from its parent
   */
  constructor(props) {
    super(props);
    this.state = {
      username: '',
      password: '',
    };
    const { match: { path } } = props;

    this.admin = path === '/admin/login';
    // bind methods
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }


  /**
   * @param {object} newProps the props that is about to be received
   * @returns {void} performs an action and returns nothing
   *
   */
  componentWillReceiveProps(newProps) {
    const { history } = this.props;

    const dashboardPath = `/dashboard${this.admin ? '/admin' : ''}`;

    if (newProps.login.success) {
      return setTimeout(() => {
        history.push(dashboardPath);
      }, 1500);
    }
  }

  /**
   * this handles all the change events that occurs in this Login
   * component
   * @param {object} e contains info about the event
   * @param {object} param1 contains the value and other properties
   * of the target
   * @returns {void} handles the event and returns nothing
   */
  handleChange(e, { name, value }) {
    this.setState({ [name]: value });
  }

  /**
   *makes an api call once the form is submitted
   * @returns {void} dispatches an action and return void
   */
  handleSubmit() {
    const { username, password } = this.state;
    const { sendLoginRequest,
      sendEngineerLoginRequest } = this.props;

    if (this.admin) {
      return sendEngineerLoginRequest({
        username, password, userType: 'engineer' });
    }
    sendLoginRequest({ username, password, userType: 'client' });
  }

  /**
   * this renders a presentation component with values gotten from state
   *@returns {JSX} returns JSX to be rendered in the DOM
   */
  render() {
    const { username, password } = this.state;
    const { login } = this.props;
    const inputs = [
      {
        name: 'username',
        placeholder: 'Username',
        value: username
      },
      {
        name: 'password',
        placeholder: 'Password',
        value: password,
        type: 'password'
      }
    ];
    const links = [];
    if (!this.admin) links.push({ caption: 'Sign Up', to: '/signup' });

    return (
      <div>
        <Header />
        <Form
          loading={login.isLoading}
          errorMessages={login.errors.errorMessages}
          messageTitle={login.errors.message || 'Login Successfully'}
          success={login.success}
          formTitle="Login"
          inputs={inputs}
          handleChange={this.handleChange}
          handleSubmit={this.handleSubmit}
          links={links}
        />
      </div>

    );
  }
}

Login.propTypes = {
  sendLoginRequest: PropTypes.func.isRequired,
  login: PropTypes.objectOf(PropTypes.any).isRequired,
  history: PropTypes.objectOf(PropTypes.any).isRequired,
  sendEngineerLoginRequest: PropTypes.func.isRequired,
  match: PropTypes.objectOf(PropTypes.any).isRequired,
};


export const mapStateToProps = ({ login }) => ({
  login,
});

export const mapDispatchToProps = dispatch => ({
  sendLoginRequest: user => dispatch(loginAction(user)),
  sendEngineerLoginRequest: engineer => dispatch(engineerLoginAction(engineer)),
});


export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Login);
