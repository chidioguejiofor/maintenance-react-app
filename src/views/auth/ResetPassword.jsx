import React, { Component } from 'react';
import Form from '../../components/Form';
import Header from '../../components/Header';
/**
 *contains logic for the LoginPage.
 */
class SignUp extends Component {
  /**
   * this creates a new instance of Login class.
   * The super constructor is called here and the state is initialized
   * @param {object} props contains the props sent from its parent
   */
  constructor(props) {
    super(props);
    this.state = {
      password: '',
      email: '',
      success: false,
    };

    // bind methods
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = () => {};
  }


  /**
   * this handles all the change events that occurs in this Login
   * component
   * @param {obejct} e contains info about the event
   * @param {object} param1 contains the value and other properties
   * of the targe
   * @returns {void} handles the event and returns nothing
   */
  handleChange(e, { name, value }) {
    this.setState({ [name]: value });
  }


  /**
   * this renders a presentation component with values gotten from state
   *@returns {JSX} returns JSX to be rendered in the DOM
   */
  render() {
    const { password, success, email } = this.state;
    const inputs = [
      {
        name: 'email',
        placeholder: 'Email',
        value: email,
        type: 'email'
      },
      {
        name: 'password',
        placeholder: 'New Password',
        value: password,
        type: 'password'
      },
      {
        name: 'confirm-password',
        placeholder: 'Confirm Password',
        value: password,
        type: 'password'
      },

    ];

    const errorMessages = [];
    const links = [
      { caption: 'Sign Up', to: '/signup' },
      { caption: 'Login', to: '/login' }
    ];
    return (
      <div>
        <Header />
        <Form
          formTitle="Reset Password"
          errorMessages={errorMessages}
          success={success}
          inputs={inputs}
          handleChange={this.handleChange}
          handleSubmit={this.handleSubmit}
          links={links}
        />
      </div>

    );
  }
}
export default SignUp;
