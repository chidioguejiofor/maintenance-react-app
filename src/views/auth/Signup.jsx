import React from 'react';


const Signup = () => (
  <h1>This is the signup page</h1>
);

export default Signup;
