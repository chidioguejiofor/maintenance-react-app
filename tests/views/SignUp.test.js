import React from 'react';
import { shallow } from 'enzyme';
import chai from 'chai';
import { SignUp, mapStateToProps, mapDispatchToProps } from '../../src/views/auth/Signup';
import AuthForm from '../../src/components/auth/AuthForm';

const resetPageSpy = jest.fn();
const getMounted = (options = {
  signup: {
    errors: { errorMessages: [] },
    data: {}
  },
  resetPage: resetPageSpy,
}) => shallow(<SignUp {...options} />);
describe('Testing SignUp Component', () => {
  describe('the components when inputs property are empty rendered in the DOM', () => {
    it('should match snapshot', () => {
      const mountedObj = getMounted();
      expect(mountedObj).toMatchSnapshot();
    });

    it('should render an AuthForm', () => {
      const mountedObj = getMounted();
      chai.expect(mountedObj.find(AuthForm).length)
        .to.equal(1);
      chai.expect(
        mountedObj.find(AuthForm)
          .prop('inputs')
          .length
      )
        .to.equal(3);
    });
  });


  describe('mapStateToProps', () => {
    it('should map the state correctly', () => {
      const mockState = {
        signup: 'mockStateSignup',
      };

      const mockedProp = mapStateToProps(mockState);

      chai.expect(mockedProp.signup)
        .to.equal(mockState.signup);
    });
  });

  describe('mapDispatchToPrps', () => {
    it('should map the state correctly', () => {
      const mockDispatch = jest.fn();

      const mockedProp = mapDispatchToProps(mockDispatch);

      chai.expect(mockedProp.sendSignUpRequest)
        .to.be.a('function');
      chai.expect(mockedProp.resetPage)
        .to.be.a('function');
      mockedProp.sendSignUpRequest();
      mockedProp.resetPage();

      expect(mockDispatch)
        .toHaveBeenCalledTimes(2);
    });
  });

  describe('the methods in the component', () => {
    describe('componentWillReceiveProps', () => {
      it('should  be called when props are updated', () => {
        const spy = jest.spyOn(SignUp.prototype, 'componentWillReceiveProps');
        const mountedObj = getMounted();
        const newProps = { temp: 'temp Props' };
        mountedObj.setProps(newProps);
        expect(spy)
          .toHaveBeenCalledTimes(1);
      });
      it('should call history.push when success equals true', () => {
        const spy = jest.spyOn(SignUp.prototype, 'componentWillReceiveProps');
        const historySpy = jest.fn();
        jest.useFakeTimers();

        const options = {
          signup: {
            errors: { errorMessages: [] },
            data: {}
          },
          history: { push: historySpy },

        };
        const mountedObj = getMounted(options);

        const newProps = {
          signup: {
            errors: { errorMessages: [] },
            success: true,
            data: {}
          },
          history: { push: historySpy },
        };
        mountedObj.setProps(newProps);
        expect(spy)
          .toHaveBeenCalledTimes(2);
        expect(setTimeout).toHaveBeenCalledTimes(1);
        jest.runAllTimers();
        expect(historySpy).toHaveBeenNthCalledWith(1, '/login');
      });


      it('should call errorMessages.push when status === 409', () => {
        const spy = jest.spyOn(SignUp.prototype, 'componentWillReceiveProps');
        const historySpy = jest.fn();
        jest.useFakeTimers();

        const options = {
          signup: {
            errors: { errorMessages: [] },
            data: {}
          },
          history: { push: historySpy },

        };
        const mountedObj = getMounted(options);

        const newProps = {
          signup: {
            errors: { errorMessages: [], statusCode: 409 },
            success: true,
            data: {}
          },
        };
        const spy2 = jest.spyOn(newProps.signup.errors.errorMessages, 'push');
        mountedObj.setProps(newProps);
        expect(spy)
          .toHaveBeenCalled();
        expect(spy2)
          .toHaveBeenNthCalledWith(1, 'Try Again');
      });
    });

    describe('componentWillUnmount', () => {
      it('should  be called when props are updated', () => {
        const spy = jest.spyOn(SignUp.prototype, 'componentWillUnmount');
        const mountedObj = getMounted();
        mountedObj.unmount();
        expect(spy)
          .toHaveBeenCalledTimes(1);
        expect(resetPageSpy)
          .toHaveBeenCalledTimes(1);
      });
    });
    describe('handleChange', () => {
      it('should update state when called', () => {
        const stateSpy = jest.spyOn(SignUp.prototype, 'setState');
        const spy = jest.spyOn(SignUp.prototype, 'handleChange');
        const mountedObj = getMounted();

        mountedObj
          .instance()
          .handleChange(null, { name: 'username', value: 'Test Username' });
        expect(spy)
          .toHaveBeenCalledTimes(1);
        expect(stateSpy)
          .toHaveBeenCalledTimes(1);
      });
    });

    describe('handleSubmit', () => {
      it('should call send login request with the an object', () => {
        const sendSignUpSpy = jest.fn();
        const spy = jest.spyOn(SignUp.prototype, 'handleChange');
        const mountedObj = getMounted({
          signup: {
            errors: { errorMessages: [] },
            data: {}
          },
          sendSignUpRequest: sendSignUpSpy,

        });

        mountedObj.instance().handleSubmit();
        expect(sendSignUpSpy)
          .toHaveBeenCalledTimes(1);
        expect(spy)
          .toHaveBeenCalledTimes(1);
      });
    });
  });
});
