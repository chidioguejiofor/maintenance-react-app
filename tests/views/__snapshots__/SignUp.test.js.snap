// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SignUp Component the components when inputs property are empty rendered in the DOM should match snapshot 1`] = `
<div>
  <AuthForm
    errorMessages={Array []}
    formType="Sign Up"
    handleChange={[Function]}
    handleSubmit={[Function]}
    inputs={
      Array [
        Object {
          "name": "email",
          "placeholder": "Email",
          "type": "email",
          "value": "",
        },
        Object {
          "name": "username",
          "placeholder": "Username",
          "value": "",
        },
        Object {
          "name": "password",
          "placeholder": "Password",
          "type": "password",
          "value": "",
        },
      ]
    }
    links={
      Array [
        Object {
          "caption": "Login",
          "to": "/login",
        },
        Object {
          "caption": "Reset Password",
          "to": "/reset-password",
        },
      ]
    }
    loading={false}
    messageTitle=""
    success={true}
  />
</div>
`;
